### ISSUE TITLE
search query evaluator is not able to get query string from Literal

### ISSUE THREAD

**Issue by jgrzebyta** [eclipse/rdf4j#220]
Affected: rdf4j version _2.0M3_
Java: 1.8
Os: Ubuntu 16.04

In Lucene search query only direct _String_ is accepted, ie: `search:query "abc123"` works fine.
In case if the term is instance of Literal taken as result of other _atomic_ request it fails. For example:

```
prefix search: <http://www.openrdf.org/contrib/lucenesail#>
prefix y: <urn:raw:yeastract#>

select ?sub ?query (isLiteral(?query) as ?isLit) ?score where {
  [] y:Yeast_id ?query .
  ?sub search:matches [
  search:query ?query ;
  search:score ?score
  ] .
}
```

I have error:

```
Exception in thread "main" org.eclipse.rdf4j.query.QueryEvaluationException: Invalid Text Query: missing query string for Lucene query specification
    at org.eclipse.rdf4j.repository.sail.SailTupleQuery.evaluate(SailTupleQuery.java:59)
    at org.eclipse.rdf4j.repository.sail.SailTupleQuery.evaluate(SailTupleQuery.java:67)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at clojure.lang.Reflector.invokeMatchingMethod(Reflector.java:93)
    at clojure.lang.Reflector.invokeInstanceMethod(Reflector.java:28)
    at sparql$process_sparql_query.invokeStatic(sparql.clj:92)
    at sparql$process_sparql_query.invoke(sparql.clj:83)
    at sparql$_main$fn__652.invoke(sparql.clj:68)
    at sparql$_main.invokeStatic(sparql.clj:67)
    at sparql$_main.doInvoke(sparql.clj:47)
    at clojure.lang.RestFn.applyTo(RestFn.java:137)
    at sparql.main(Unknown Source)
Caused by: org.eclipse.rdf4j.sail.SailException: Invalid Text Query: missing query string for Lucene query specification
    at org.eclipse.rdf4j.sail.lucene.QuerySpecBuilder.failOrWarn(QuerySpecBuilder.java:203)
    at org.eclipse.rdf4j.sail.lucene.QuerySpecBuilder.process(QuerySpecBuilder.java:137)
    at org.eclipse.rdf4j.sail.lucene.LuceneSailConnection.evaluate(LuceneSailConnection.java:294)
    at org.eclipse.rdf4j.repository.sail.SailTupleQuery.evaluate(SailTupleQuery.java:50)
    ... 14 more
```

`?query` is literal:

```
prefix search: <http://www.openrdf.org/contrib/lucenesail#>
prefix y: <urn:raw:yeastract#>

select ?query (isLiteral(?query) as ?isLit) where {
  [] y:Yeast_id ?query .

}
```

Gives:

```
query,isLit
Abf1,true
Abf2,true
Arr1,true
Put3,true
Rap1,true
```

**Comment by jgrzebyta** [comment]
Maybe it is important:

To get query result I do:

1. Instantiate a query by `RepositoryConnection.prepareQuery`
2. Instantiate writer based on query type. For that example it is `SPARQLResultsCSVWriter(System.out)`
3. Execute the query with `query.evaluate(writer)`

Probably this is the reason that in the very bottom example `?query` values are not URI. They should be serialised with `<http://www.w3.org/2001/XMLSchema#string>` type.

**Comment by jgrzebyta** [comment]
I have reproduced this error in my project. 
The datasets are in files: 
- `tests/resources/yeastract_raw.ttl`
- `tests/resources/Uniprot_scer.rdf`

sparql query is in:
- `tests/resouces/issue220.sparql`

**Comment by pulquero** [comment]
I believe this is a known limitation of the current search impl. The search is always executed first and then joined with the rest of the query. So in your case ?query is not bound when the search executes. It could probably be fixed by wrapping SearchQueryEvaluator in a TupleFunction. A workaround would be to split the query in two, execute the first part then using the result bindings execute the second part.

**Comment by jgrzebyta** [comment]
I have started working on this and #235 issues.

**Comment by pulquero** [comment]
ok, if you need any help let me know.

**Comment by jgrzebyta** [comment]
@pulquero  I have built a mock tests which emulates new syntax of search:match feature (commit f5cda4e40199104326ef883680728a6d161b8101). The new implementation will meet syntax described in https://jena.apache.org/documentation/query/text-query.html

I found there is a major (imho) issues which should be found: Do RDF4J uses any mechanism of bean's injection, eg passing `LuceneIndex` into the searching engine? If not I would use the Thread binding.
